### ISSUE TITLE
Inversion of the starting model

### ISSUE THREAD

**Issue by ChangJingRun1997** [gimli-org/gimli#379]
### Problem description
Hello,I am working in an unbounded(half-space) domain.
My starting model is divided into 2 areas, how do I set the resistivity of the background and the resistivity of area 2 according to the markers.Mark 1 is background resistivity of 5 ohm, Mark 2 is 1000 ohm

### Your environment

**Operating system**:  Mac
**Python version**: 3.10
**pyGIMLi version**: 1.2.4
### Steps to reproduce

Tell us how to reproduce this issue. Ideally, you could paste the code that produces the error:

```python
import pygimli.meshtools as mt
from pygimli.physics import ert
cube = mt.createCube(size=[100, 15, 20], pos=[0, 0, -10], boundaryMarker=1,marker=1)
cube_hdpe=mt.createCube(size=[96,0.1,10],pos=[0,0,-5],marker=2)
plc=cube+cube_hdpe
plc.exportVTK('plc')
data=ert.load("data.dat")
for s in data.sensors():
      plc.createNode(s, marker=-99)
      #plc.createNode(s - 1e-1 / 2)
plc.createNode([48,-5,0],marker=-1000)
mesh = mt.createMesh(plc, area=1,quality=34)
#pg.show(mesh,mesh.cellMarkers())
mesh.exportVTK("mesh.vtk")
rhomap=[[1,5],[2,1000]]
data["err"] = ert.estimateError(data)
mgr=ert.ERTManager()
inv=mgr.invert(data=data,mesh=mesh,lam=10,verbose=True,startModel=[])
mgr.saveResult()
```
WARNING - Region Nr: 1  is background and should not get a model transformation.

**Comment by halbmy** [comment]
In ERT in unbounded domains, we typically have an inner mesh where the inversion is done, and an outer mesh that is used for the forward calculation in order to diminish the influence of the boundary conditions. This is why by default the region with the lowest marker is taken as background region (i.e. it is not part of inversion but will contain prolongated conductivity in the forward calculation).

If the inversion region is big enough you can work without an outer region, but you will have to unset the background behaviour by `mgr.fop.setRegionProperties(1, background=False)`.

**Comment by ChangJingRun1997** [comment]
Thanks for the suggestion, you mean that the background field does not change during the inversion. If so, that's good, my concern is the change of resistivity of my marker 2. How do I pass the marker to give the starting model before inversion? I think definitely need to give the resistivity of the background field and the resistivity of marker 2 in order to do the calculations for the forward.
It's too bad that the inversion is still not working properly.

**Comment by halbmy** [comment]
A `background` region is not part of the inversion, but is automatically filled from adjacent regions for the forward calculations. There is another property called `single` that just consists of one unknown in the inversion no matter how many cells there are in. Of course you can also set the starting or reference model or the model limits.

**Comment by ChangJingRun1997** [comment]
`import pygimli.meshtools as mt
from pygimli.physics import ert
cube_region_1=mt.createCube(size=[150,15,60],pos=[0,0,-30],boundaryMarker=True,marker=1)
cube_region_2=mt.createCube(size=[100, 10, 40], pos=[0, 0, -20], marker=2)
cube_region_3=mt.createCube(size=[96, 0.5, 30],pos=[0, 0, -15], marker=3)
plc=cube_region_1+cube_region_2+cube_region_3
plc.exportVTK('plc')
data = ert.load("data.dat")
data["err"] = ert.estimateError(data)
for s in data.sensors():
    plc.createNode(s)
    plc.createNode(s - 1e-1 / 2)
mesh = mt.createMesh(plc,area=10,quality=34)
#mesh.exportVTK("mesh.vtk")
mgr=ert.ERTManager()
mgr.setMesh(mesh)
mgr.setData(data)
mgr.fop.region(2).setStartValue(5)
mgr.fop.region(2).setSingle(True)
mgr.fop.region(3).setStartValue(1000000)
mgr.invert(startModelIsReference=True,lam=10,zWeight=0.1,verbose=True)
mgr.saveResult()
[Uploading data.dat…]()
`

**Comment by ChangJingRun1997** [comment]
`
import pygimli.meshtools as mt
from pygimli.physics import ert
cube_region_1=mt.createCube(size=[150,15,60],pos=[0,0,-30],boundaryMarker=True, marker=1)
cube_region_2=mt.createCube(size=[100, 10, 40], pos=[0, 0, -20], marker=2)
cube_region_3=mt.createCube(size=[96, 0.5, 30],pos=[0, 0, -15], marker=3)
plc=cube_region_1+cube_region_2+cube_region_3
#plc.exportVTK('plc')
data = ert.load("data.dat")
data["err"] = ert.estimateError(data)
for s in data.sensors():
    plc.createNode(s)
    plc.createNode(s - 1e-1 / 2)
mesh = mt.createMesh(plc,area=10,quality=34)
#mesh.exportVTK("mesh.vtk")
mgr=ert.ERTManager()
mgr.setMesh(mesh)
mgr.setData(data)
mgr.fop.region(2).setStartValue(5)
mgr.fop.region(2).setSingle(True)
mgr.fop.region(3).setStartValue(1000000)
mgr.invert(startModelIsReference=True, lam=10,zWeight=0.1, verbose=True)
mgr.saveResult()
`
[data.zip](https://github.com/gimli-org/gimli/files/8649754/data.zip)

Based on your advice, I am not sure where the error is now

**Comment by halbmy** [comment]
Actually, acting on the region classes itself is outdated and will be disabled in the future. Since pg 1.2.5 we prefer the following style (see also https://github.com/gimli-org/transform2022/blob/main/2_Lake.ipynb):
```
mgr.inv.setRegularization(single=True)  # for the whole domain
```
or 
```
mgr.inv.setRegularization(3, fix=1000)  # for one region only
```

**Comment by ChangJingRun1997** [comment]
`mgr=ert.ERTManager(data, verbose=True)
mgr.setMesh(mesh)
mgr.inv.setRegularization(2, startModel=5, single=True)
mgr.inv.setRegularization(3, startModel=1000000)  
mgr.invert(verbose=True)
`
It still comes up with errors
RuntimeError: /opt/anaconda3/conda-bld/pgcore_1645176476263/work/gimli/core/src/bert/bertJacobian.cpp:289		void GIMLI::createSensitivityCol_(Matrix<ValueType> &, const GIMLI::Mesh &, const GIMLI::DataContainerERT &, const Matrix<ValueType> &, const GIMLI::RVector &, const GIMLI::RVector &, std::vector<std::pair<Index, Index>> &, GIMLI::uint, bool) [ValueType = double]  potential matrix rowsize to small.127 < 128

Thank you again for your reply！

**Comment by halbmy** [comment]
A few comments:
1. a mesh quality of 34 does not make sense for 3D meshes. Use 1.3 or similar (see tetgen)
2. if you want to model a void with your region 3, consider using a void (`createCube(isHole=True`) or fixing the resistivity (`inv.setRegularization(3, fix=1e7)`)
3. the mesh needs to be much bigger to account for boundary effects, region 1, but also region 2
4. set region 1 explicitly to background (`inv.setRegularization(1, background=True)`)
5. I doubt that you can resolve much with pole-pole data
6. The data do not really allow for 3D inversion, maybe a 2D inverse model (in a 3D forward model) makes more sense

**Comment by ChangJingRun1997** [comment]
Thank you for your comments, they are very valuable to me!
1.About the data
         top view
![演示文稿1](https://user-images.githubusercontent.com/99867753/167646444-d6561d03-0558-48d3-9685-c7840db17c4f.png)
AM(pole-pole)
#x y z
-31 1 0 #A1
-31 2 0 #A2
...          #Mn
#a b m n
1 0 65 0
...
Do we still need to give the coordinates of the electrodes in the infinite distance?
2. About the region
2.1Region_1 as background
`inv.setRegularization(1, background=True)`
2.2The region 2 is a fixed value
`inv.setRegularization(2,fix=5)`
2.3The change in resistivity of region 3 is what I focus on, I need to give it an initial resistivity of 1e6. The resistivity of region 3 cannot be fixed to a single value
`inv.setRegularization(3,startModel=1e6)`
It still doesn't work properly.

**Comment by halbmy** [comment]
1. No a zero is for remote electrodes.
2. Is the resistivity in region 2 really constantly 5 Ohmm? If not, better leave it open by `single=True`

If the inside resistivity is several orders of magnitude higher than the outside one, the measured voltage will not depend much from the actual value as you are in saturation, I am afraid. Some modelling should be done before.

**Comment by halbmy** [comment]
What exactly does not work? The same error message as above?

**Comment by ChangJingRun1997** [comment]
Yes, the program and data are below, and I would like to get more comments from you.
[test.zip](https://github.com/gimli-org/gimli/files/8663052/test.zip)

**Comment by halbmy** [comment]
The data look completely different from the one in the BERT issue 269
```
pg.viewer.mpl.showDataContainerAsMatrix(data, "a", "m", "rhoa", cMap="Spectral_r")
```

![image](https://user-images.githubusercontent.com/7136401/172899609-e89652f5-844c-4a65-82a1-52d92fe72d61.png)

I still can't understand these very low values. Can you produce such an image using a synthetic model using your idea of the region resistivity?

**Comment by halbmy** [comment]
The data look completely different from the one in the BERT issue 269
```
pg.viewer.mpl.showDataContainerAsMatrix(data, "a", "m", "rhoa", cMap="Spectral_r")
```

![image](https://user-images.githubusercontent.com/7136401/172899609-e89652f5-844c-4a65-82a1-52d92fe72d61.png)

I still can't understand these very low values. Can you produce such an image using a synthetic model using your idea of the region resistivity?

**Comment by halbmy** [comment]
The data look completely different from the one in the BERT issue 269
```
pg.viewer.mpl.showDataContainerAsMatrix(data, "a", "m", "rhoa", cMap="Spectral_r")
```

![image](https://user-images.githubusercontent.com/7136401/172899609-e89652f5-844c-4a65-82a1-52d92fe72d61.png)

I still can't understand these very low values. Can you produce such an image using a synthetic model using your idea of the region resistivity?

**Comment by ChangJingRun1997** [comment]
When there is high resistance in the middle, the potential value of M is very small.
`data["rhoa"] = data["u"] / data["i"] * data["k"]`
Can you provide some examples of the synthesis model?
Thank you!

**Comment by halbmy** [comment]
I don't understand the first part. Have a look at pygimli.org for examples on synthetic ERT modelling and mesh generation.

When working in an infinite halfspace, better use `mt.createWorld` instead of `mt.createCube` because the first will set correct boundary conditions (Neumann on top and mixed in the subsurface).

**Comment by ChangJingRun1997** [comment]
I want to simulate the internal changes of high resistance by the size of the resistivity. I don't know what is wrong with the code, which is completely different from the COMSOL calculation.
`import pygimli.meshtools as mt
from pygimli.physics import ert
import pygimli as pg
cube_region_1=mt.createWorld(start=[-50,-15,-40],end=[150,15,0],worldMarker=True,marker=1)
cube_region_2=mt.createCube(size=[100,0.4,30],pos=[50, 0, -15], marker=2)
plc=cube_region_1+cube_region_2
shm=pg.DataContainerERT('data1.dat')
for s in shm.sensors():
    plc.createNode(s)
for s in shm.sensorPositions():
    plc.createNode(s - [0,1e-1/2,0])
mesh = mt.createMesh(plc,quality=1.4)
res = [[1, 5.0], [2, 1e6]]
data = ert.simulate(mesh, res=res, scheme=shm, verbose=True)
data.remove(data['rhoa'] < 0)
pg.viewer.mpl.showDataContainerAsMatrix(data, "a", "m", "rhoa", cMap="jet")
`
<img width="525" alt="image" src="https://user-images.githubusercontent.com/99867753/177267585-7584d10a-92fa-42a8-9c9b-856f4bd2ed9a.png">
`
cube_region_1=mt.createWorld(start=[-50,-15,-40],end=[150,15,0],worldMarker=True,marker=1)
cube_region_2=mt.createCube(size=[100,0.4,30],pos=[50, 0, -15], marker=2)
cube_region_3=mt.createCube(size=[1,0.4,1],pos=[50, 0, -3],marker=3)
plc=cube_region_1+cube_region_2+cube_region_3
shm=pg.DataContainerERT('data1.dat')
for s in shm.sensors():
    plc.createNode(s)
for s in shm.sensorPositions():
    plc.createNode(s - [0,1e-1/2,0])
mesh = mt.createMesh(plc,quality=1.4)
res = [[1, 5.0], [2, 1e6],[3,2]]
data = ert.simulate(mesh, res=res, scheme=shm, verbose=True)
data.remove(data['rhoa'] < 0)
pg.viewer.mpl.showDataContainerAsMatrix(data, "a", "m", "rhoa", cMap="jet")
`
[Uploading data1.dat.zip…]()

<img width="554" alt="image" src="https://user-images.githubusercontent.com/99867753/177268442-416f16b4-b5fb-4d3e-abe0-914752e342b9.png">
marker 3 doesn't do much, unlike COMSOL.

**Comment by ChangJingRun1997** [comment]
Thank you for making some progress recently.
How do I control the grid size of region 2, now it's too dense. I want the grid size to get bigger and bigger from the surface down to the depths.
`cube = mt.createCube(size=[100, 15, 20], pos=[0, 0, -10], boundaryMarker=1,marker=1)
cube_2=mt.createCube(size=[96,0.1,10],pos=[0,0,-5],marker=2)
plc=cube+cube_2
mesh = mt.createMesh(plc,quality=1.4)`
area=，Has little effect on the grid size of region 2

**Comment by halbmy** [comment]
The problem is that the inner cube is just 10cm wide and therefore the tetrahedra are also 0.1cm small. 

In this case it could make more sense to use a structured rectangular (hexahedra) grid using `mt.createGrid`.

**Comment by ChangJingRun1997** [comment]
`grid=mt.createGrid(x=x, y=y, z=z, woldBoundaryMarker=True)
for c in grid.cells():
     c.setMarker(1)
....
res=[[1,5],[2,1e6]]
data=ert.simulate(grid,res=res,...)
`

**Comment by halbmy** [comment]
You will have to loop over the cells:
```
for c in grid.cells():
    if c.center().x() < xmin:
        c.setMarker(2)
...
```
or
```
cm = grid.cellMarkers()
cx = pg.x(grid.cellCenters())
cm[cx < xmin] = 2
...
```
